users/vieting/experiments/switchboard/ctc/feat/network_helpers/specaug_configurable.py

```diff
@@ -223,5 +302,152 @@ def logging_ops():
+        freq_mask_max_num = tf.gather(specaug_params["freq_mask_max_num"], current_epoch)
+        freq_mask_max_size = tf.gather(specaug_params["freq_mask_max_size"], current_epoch)
+        total_time_masks_max_frames = tf.cast(
+            tf.math.floor(config["time_mask_max_proportion"] * tf.cast(tf.shape(x)[data.time_dim_axis], tf.float32)),
+            tf.int32,
+        )
+        total_freq_masks_max_size = tf.cast(
+            tf.math.floor(config["freq_mask_max_proportion"] * tf.cast(tf.shape(x)[data.feature_dim_axis], tf.float32)),
+            tf.int32,
+        )
+        max_time_num_seq_len = tf.cast(
+            tf.math.floordiv(
+                tf.cast(tf.shape(x)[data.time_dim_axis], tf.float32),
+                tf.cast(1.0, tf.float32) / max_time_num_seq_len_divisor * tf.cast(time_mask_max_size, tf.float32),
+            ),
+            tf.int32,
+        )
+        # check for the limits
+        actual_time_mask_max_num = tf.minimum(
+            tf.maximum(
+                time_mask_max_num,
+                max_time_num_seq_len,
+            ),
+            total_time_masks_max_frames // time_mask_max_size,
+        )
+        actual_freq_mask_max_num = tf.minimum(freq_mask_max_num, total_freq_masks_max_size // freq_mask_max_size)
+
+        if config["filter_based_masking_strategy"] == "variance":
+            f_resp = get_frequency_response(filter_layer)
+            variance = tf.math.reduce_variance(f_resp, axis=0)
+            n_features = tf.shape(x)[data.feature_dim_axis]
+            probs = variance / tf.reduce_sum(variance)
+            uniform_probs = tf.ones_like(probs) / tf.cast(n_features, tf.float32)
+            final_probs = filter_factor * probs + (1 - filter_factor) * uniform_probs
+        elif config["filter_based_masking_strategy"] == "peakToAverage":
+            # Get peak to average ratio for each filter
+            f_resp = get_frequency_response(filter_layer)
+            peak = tf.reduce_max(f_resp, axis=0)
+            average = tf.reduce_mean(f_resp, axis=0)
+            ratio = peak / average
+            n_features = tf.shape(x)[data.feature_dim_axis]
+            # Normalize the ratio to get probabilities
+            probs = ratio / tf.reduce_sum(ratio)
+            uniform_probs = tf.ones_like(probs) / tf.cast(n_features, tf.float32)
+            final_probs = filter_factor * probs + (1 - filter_factor) * uniform_probs
+
+        enable_logging = tf.convert_to_tensor(config["enable_logging"], dtype=tf.bool)
+
+        def logging_ops():
+            with tf.control_dependencies(
+                [
+                    tf.print(
+                        "Specaug Log: ",
+                        current_epoch,
+                        tf.shape(x)[data.time_dim_axis],
+                        sep=", ",
+                    )
+                ]
+            ):
+                return tf.identity(x_masked)
```

Comment:
The `logging_ops` method is defined twice if I see it correctly. Also, it does not really contain that much useful information, does it?

Reply:
Yes. I defined it twice in the transform functions so that I do not have to import it into the config as well and and change more hashes. The log does not contain much information, because most info about the mask size should be reported from inside the `_mask` function where the size is known. This type of logging will be added in a future PR.

Reply:
I don't understand how defining it twice here is related to hashes. Adding more detailed logging later seems alright.

Reply:
I see the problem. That was a merge artifact. I removed one of them